File: serializer.go
package main

import (
	"encoding/json"
	"strings"
)

type Serializeable interface{}

func Deserialize(data []byte, s Serializeable) error {
	return json.Unmarshal(data, s)
}

func Serialize(s Serializeable) ([]byte, error) {
	return json.Marshal(s)
}

func extractMessagesFromBuffer(buffer string) ([]Message, string, error) {
	var messages []Message

	validBuffer := buffer[:strings.LastIndex(buffer, "}")+1]
	invalidBuffer := buffer[strings.LastIndex(buffer, "}")+1:]

	// Debugging...

	var err error
	for _, msg := range strings.Split(validBuffer, "}{") {
		// We need to add the closing bracket back to the message

		if !strings.HasSuffix(msg, "}") {
			msg += "}"
		}

		var message Message
		err = Deserialize([]byte(msg), &message)
		if err != nil {
			break
		}

		messages = append(messages, message)
	}

	return messages, invalidBuffer, err
}

File: task/serializer.py
import json
import pyperclip
from model import Message

def serialize(obj):
    return json.dumps(obj.to_dict()).encode()

def deserialize(data, cls):
    return cls.from_dict(json.loads(data))

def extract_messages_from_buffer(buffer):
    buffer_str = buffer.decode()
    valid_buffer = buffer_str[:buffer_str.rfind("}") + 1]
    invalid_buffer = buffer_str[buffer_str.rfind("}") + 1:]
    
    pyperclip.copy(valid_buffer)
    
    messages = []
    for msg_str in valid_buffer.split("}{"):
        msg_str = msg_str if msg_str.endswith("}") else msg_str + "}"
        msg = deserialize(msg_str.encode(), Message)
        messages.append(msg)
    
    return messages, invalid_buffer.encode()
